### PR TITLE
std: Move overflowing ops to inherent methods

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit migrates all of the methods on `num::wrapping::OverflowingOps` onto
inherent methods of the integer types. This also fills out some missing gaps in
the saturating and checked departments such as:
- `saturating_mul`
- `checked_{neg,rem,shl,shr}`

This is done in preparation for stabilization,

cc #27755

### CODE REVIEW

**Comment by alexcrichton** [comment]
r? @aturon 

Note that this doesn't deprecate the old wrappers, it just prepares everything for the upcoming "deprecation commit"

**Comment by aturon** [comment]
Hm, this might be slightly premature given discussion [here](https://github.com/rust-lang/rust/issues/27755#issuecomment-165628303). Let's wait for some overall consensus before landing this change.

File: src/libcore/num/mod.rs [hunk 1]
```diff
@@ -662,6 +783,225 @@ macro_rules! int_impl {
...
+                (0, true)
+            } else {
+                (self % rhs, false)
+            }
+        }
+
+        /// Negates self, overflowing if this is equal to the minimum value.
+        ///
+        /// Returns a tuple of the negated version of self along with a boolean
+        /// indicating whether an overflow happened.
```
Comment (aturon) [inline src/libcore/num/mod.rs]: This comment isn't quite clear about what value is returned on overflow.

File: src/libcore/num/mod.rs [hunk 2]
```diff
@@ -662,6 +783,225 @@ macro_rules! int_impl {
...
+                (-self, false)
+            }
+        }
+
+        /// Shifts self left by `rhs` bits.
+        ///
+        /// Returns a tuple of the shifted version of self along with a boolean
+        /// indicating whether the shift value was larger than or equal to the
+        /// number of bits. If the shift value is too large, then only the lower
+        /// bits of the shift value are used to perform a shift.
```
Comment (aturon) [inline src/libcore/num/mod.rs]: "Only the lower bits" is a bit vague, can you make this more clear?

File: src/libcore/num/mod.rs [hunk 3]
```diff
@@ -662,6 +783,225 @@ macro_rules! int_impl {
...
+        #[unstable(feature = "wrapping", issue = "27755")]
+        pub fn overflowing_shl(self, rhs: u32) -> (Self, bool) {
+            (self << (rhs & ($BITS - 1)), (rhs > ($BITS - 1)))
+        }
+
+        /// Shifts self right by `rhs` bits.
+        ///
+        /// Returns a tuple of the shifted version of self along with a boolean
+        /// indicating whether the shift value was larger than or equal to the
+        /// number of bits. If the shift value is too large, then only the lower
```
Comment (aturon) [inline src/libcore/num/mod.rs]: "Only the lower bits" is a bit vague, can you make this more clear?

File: src/libcore/num/mod.rs [hunk 4]
```diff
@@ -1464,6 +1891,208 @@ macro_rules! uint_impl {
...
+        pub fn overflowing_div(self, rhs: Self) -> (Self, bool) {
+            (self / rhs, false)
+        }
+
+        /// Calculates the remainder when `self` is divided by `rhs`.
+        ///
+        /// Returns a tuple of the remainder after dividing along with a boolean
+        /// indicating whether an arithmetic overflow would occur. Note that for
+        /// unsigned integers overflow never occurs, so the second value is
+        /// always `false`.
```
Comment (aturon) [inline src/libcore/num/mod.rs]: Hmm... does it even make sense to provide this method?

File: src/libcore/num/mod.rs [hunk 5]
```diff
@@ -1464,6 +1891,208 @@ macro_rules! uint_impl {
...
+                (a as Self, b)
+            }
+        }
+
+        /// Calculates the divisor when `self` is divided by `rhs`.
+        ///
+        /// Returns a tuple of the divisor along with a boolean indicating
+        /// whether an arithmetic overflow would occur. Note that for unsigned
+        /// integers overflow never occurs, so the second value is always
+        /// `false`.
```
Comment (aturon) [inline src/libcore/num/mod.rs]: Hmm... does it even make sense to provide this method?

File: src/libcore/num/mod.rs [hunk 6]
```diff
@@ -1464,6 +1891,208 @@ macro_rules! uint_impl {
...
+        #[unstable(feature = "wrapping", issue = "27755")]
+        pub fn overflowing_rem(self, rhs: Self) -> (Self, bool) {
+            (self % rhs, false)
+        }
+
+        /// Negates self in an overflowing fashion.
+        ///
+        /// Returns `!self + 1` using wrapping operations to return the value
+        /// that represents the negation of this unsigned value. Note that for
+        /// unsigned values overflow always occurs, so `true` is always returned
```
Comment (aturon) [inline src/libcore/num/mod.rs]: Same as earlier comment.

File: src/libcore/num/mod.rs [hunk 7]
```diff
@@ -1464,6 +1891,208 @@ macro_rules! uint_impl {
...
+        pub fn overflowing_neg(self) -> (Self, bool) {
+            ((!self).wrapping_add(1), true)
+        }
+
+        /// Shifts self left by `rhs` bits.
+        ///
+        /// Returns a tuple of the shifted version of self along with a boolean
+        /// indicating whether the shift value was larger than or equal to the
+        /// number of bits. If the shift value is too large, then only the lower
+        /// bits of the shift value are used to perform a shift.
```
Comment (aturon) [inline src/libcore/num/mod.rs]: Same as earlier comment.

File: src/libcore/num/mod.rs [hunk 8]
```diff
@@ -1464,6 +1891,208 @@ macro_rules! uint_impl {
...
+        pub fn overflowing_shl(self, rhs: u32) -> (Self, bool) {
+            (self << (rhs & ($BITS - 1)), (rhs > ($BITS - 1)))
+        }
+
+        /// Shifts self right by `rhs` bits.
+        ///
+        /// Returns a tuple of the shifted version of self along with a boolean
+        /// indicating whether the shift value was larger than or equal to the
+        /// number of bits. If the shift value is too large, then only the lower
+        /// bits of the shift value are used to perform a shift.
```
Comment (aturon) [inline src/libcore/num/mod.rs]: Same as earlier comment.

**Comment by aturon** [comment]
@alexcrichton There seems to be consensus around the issue, so I've reviewed the PR and left a few nits. Once these are resolved r=me.

**Comment by ollie27** [comment]
I wouldn't expect `overflowing_neg` to return `true` when given `0` for unsigned ints as `-0` can be represented as an unsigned int: `0`.

Also it looks like `checked_neg` is missing for unsigned ints.

File: src/libcore/num/mod.rs (hunk 4 above)
Comment (alexcrichton) [inline src/libcore/num/mod.rs]: I figured it made sense to do so in the face of consistency with unsigned integers. This means you can just call the method blindly and not worry about which type lies underneath.

**Comment by alexcrichton** [comment]
@ollie27 ah good point about negating 0!

@aturon I've updated with hopefully a clarification about shifting and overflow with the mask, although it may still be a little unclear as it's somewhat technical, what do you think?

**Pull request merged**